crypto/cmp/cmp_asn.c

```diff
@@ -188,6 +235,361 @@ int OSSL_CMP_ITAV_push0_stack_item(STACK_OF(OSSL_CMP_ITAV) **itav_sk_p,
+    if (itav == NULL || OBJ_obj2nid(itav->infoType) != NID_id_it_caCerts) {
+        ERR_raise(ERR_LIB_CMP, ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+    *out = itav->infoValue.caCerts;
+    return 1;
+}
+
+OSSL_CMP_ITAV *OSSL_CMP_ITAV_new0_rootCaKeyUpdate(X509 *newWithNew,
+                                                  X509 *newWithOld,
+                                                  X509 *oldWithNew)
+{
+    OSSL_CMP_ITAV *itav;
+    OSSL_CMP_ROOTCAKEYUPDATE *upd;
+
+    if ((upd = OSSL_CMP_ROOTCAKEYUPDATE_new()) == NULL)
+        return NULL;
+    if ((itav = OSSL_CMP_ITAV_new()) == NULL) {
+        OSSL_CMP_ROOTCAKEYUPDATE_free(upd);
+        return NULL;
+    }
+    itav->infoType = OBJ_nid2obj(NID_id_it_rootCaKeyUpdate);
+    itav->infoValue.rootCaKeyUpdate = upd;
+    upd->newWithNew = newWithNew;
+    upd->newWithOld = newWithOld;
+    upd->oldWithNew = oldWithNew;
+    return itav;
+}
+
+int OSSL_CMP_ITAV_get0_rootCaKeyUpdate(const OSSL_CMP_ITAV *itav,
+                                       X509 **newWithNew,
+                                       X509 **newWithOld,
+                                       X509 **oldWithNew)
+{
+    OSSL_CMP_ROOTCAKEYUPDATE *upd;
+
+    if (newWithNew == NULL) {
+        ERR_raise(ERR_LIB_CMP, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    *newWithNew = NULL;
+    if (newWithOld != NULL)
+        *newWithOld = NULL;
+    if (oldWithNew != NULL)
+        *oldWithNew = NULL;
+    if (itav == NULL
+            || OBJ_obj2nid(itav->infoType) != NID_id_it_rootCaKeyUpdate) {
+        ERR_raise(ERR_LIB_CMP, ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+    upd = itav->infoValue.rootCaKeyUpdate;
+    *newWithNew = upd->newWithNew;
+    if (newWithOld != NULL)
+        *newWithOld = upd->newWithOld;
+    if (oldWithNew != NULL)
+        *oldWithNew = upd->oldWithNew;
+    return 1;
+}
+
+OSSL_CMP_ITAV
+*OSSL_CMP_ITAV_new0_certReqTemplate(OSSL_CRMF_CERTTEMPLATE *certTemplate,
+                                    STACK_OF(OSSL_CMP_ATAV) *keySpec)
+{
+    OSSL_CMP_ITAV *itav;
+    OSSL_CMP_CERTREQTEMPLATE *tmpl;
+
+    if ((tmpl = OSSL_CMP_CERTREQTEMPLATE_new()) == NULL)
+        return NULL;
+    if ((itav = OSSL_CMP_ITAV_new()) == NULL) {
+        OSSL_CMP_CERTREQTEMPLATE_free(tmpl);
+        return NULL;
+    }
+    itav->infoType = OBJ_nid2obj(NID_id_it_certReqTemplate);
+    itav->infoValue.certReqTemplate = tmpl;
+    tmpl->certTemplate = certTemplate;
+    tmpl->keySpec = keySpec;
+    return itav;
+}
+
+int OSSL_CMP_ITAV_get0_certReqTemplate(const OSSL_CMP_ITAV *itav,
+                                       OSSL_CRMF_CERTTEMPLATE **certTemplate,
+                                       STACK_OF(OSSL_CMP_ATAV) **keySpec)
+{
+    OSSL_CMP_CERTREQTEMPLATE *req;
+
+    if (certTemplate == NULL) {
+        ERR_raise(ERR_LIB_CMP, CMP_R_NULL_ARGUMENT);
+        return 0;
+    }
+    *certTemplate = NULL;
+    if (keySpec != NULL)
+        *keySpec = NULL;
```

Comment:
Is this necessary here? IMO it is better to not touch the output argument in case of error.

Reply:
I've removed this kind of initialization here and also in all other occurrences mentioned below.